File: macos/Tests/JasnostCaptureCoreTests/OtlpMapperTests.swift
import XCTest

@testable import JasnostCaptureCore

/// Deterministic RNG so trace/span ids are assertable.
private struct CountingRNG: RandomNumberGenerator {
    var state: UInt64 = 0
    mutating func next() -> UInt64 {
        state += 1
        return state
    }
}

private func attributeDict(_ attrs: [Otlp.KeyValue]) -> [String: Otlp.AnyValue] {
    var dict: [String: Otlp.AnyValue] = [:]
    for kv in attrs {
        XCTAssertNil(dict[kv.key], "duplicate attribute key \(kv.key)")
        dict[kv.key] = kv.value
    }
    return dict
}

/// Tests for the OTLP/JSON wire models — value encoding must match what the live Keboola
/// Data Stream accepted (intValue as STRING, doubleValue/boolValue native).
final class OtlpModelTests: XCTestCase {
    func testAnyValueEncodings() throws {
        let cases: [(Otlp.AnyValue, String)] = [
            (.string("x"), #"{"stringValue":"x"}"#),
            (.bool(false), #"{"boolValue":false}"#),
            (.int(123), #"{"intValue":"123"}"#),  // proto3 JSON: int64 is a decimal string
            (.double(1.5), #"{"doubleValue":1.5}"#),
        ]
        for (value, expected) in cases {
            let data = try JSONEncoder().encode(value)
            XCTAssertEqual(String(decoding: data, as: UTF8.self), expected)
        }
    }

    func testAnyValueDecodingRoundtripAndLenientInt() throws {
        for fixture in [#"{"stringValue":"x"}"#, #"{"boolValue":true}"#, #"{"doubleValue":2.5}"#] {
            let decoded = try JSONDecoder().decode(Otlp.AnyValue.self, from: Data(fixture.utf8))
            let reencoded = try JSONEncoder().encode(decoded)
            XCTAssertEqual(String(decoding: reencoded, as: UTF8.self), fixture)
        }
        // Spec string form and lenient bare-number form both decode to .int.
        for fixture in [#"{"intValue":"42"}"#, #"{"intValue":42}"#] {
            let decoded = try JSONDecoder().decode(Otlp.AnyValue.self, from: Data(fixture.utf8))
            XCTAssertEqual(decoded, .int(42))
        }
        // An empty object is not a valid AnyValue.
        XCTAssertThrowsError(try JSONDecoder().decode(Otlp.AnyValue.self, from: Data("{}".utf8)))
    }

    /// Decode a trimmed copy of the LIVE-ACCEPTED payload (HTTP 200 from the real stream,
    /// 2026-06-13) — proves the Codable models match the verified wire shape.
    func testDecodesLiveVerifiedLogsFixture() throws {
        let fixture = """
            {"resourceLogs":[{"resource":{"attributes":[
              {"key":"service.name","value":{"stringValue":"jasnost-capture"}},
              {"key":"enduser.id","value":{"stringValue":"petr@example.com"}}]},
             "scopeLogs":[{"scope":{"name":"jasnost.agent"},"logRecords":[{
               "timeUnixNano":"1781305072000000000",
               "observedTimeUnixNano":"1781305072000000000",
               "severityText":"INFO","severityNumber":9,
               "traceId":"aaaa1111bbbb2222cccc3333dddd4444","spanId":"abcd1234abcd1234",
               "body":{"stringValue":"click"},
               "attributes":[
                 {"key":"sequence","value":{"intValue":"1"}},
                 {"key":"input_masked","value":{"boolValue":false}},
                 {"key":"target.boundingBox.x","value":{"doubleValue":100.5}}]}]}]}]}
            """
        let request = try JSONDecoder().decode(
            Otlp.ExportLogsServiceRequest.self, from: Data(fixture.utf8))
        let record = request.resourceLogs[0].scopeLogs[0].logRecords[0]
        XCTAssertEqual(record.body, .string("click"))
        XCTAssertEqual(record.traceId, "aaaa1111bbbb2222cccc3333dddd4444")
        XCTAssertEqual(
            attributeDict(record.attributes),
            [
                "sequence": .int(1),
                "input_masked": .bool(false),
                "target.boundingBox.x": .double(100.5),
            ])
    }
}

final class OtlpIdsTests: XCTestCase {
    func testInjectableRNGProducesDeterministicIds() {
        var rng = CountingRNG()
        XCTAssertEqual(OtlpIds.traceId(using: &rng), "00000000000000010000000000000002")
        XCTAssertEqual(OtlpIds.spanId(using: &rng), "0000000000000003")
    }

    func testSystemRNGShapes() {
        let trace = OtlpIds.traceId()
        let span = OtlpIds.spanId()
        XCTAssertEqual(trace.count, 32)
        XCTAssertEqual(span.count, 16)
        XCTAssertTrue(trace.allSatisfy(\.isHexDigit))
        XCTAssertTrue(span.allSatisfy(\.isHexDigit))
        XCTAssertNotEqual(OtlpIds.traceId(), trace, "two random trace ids should differ")
    }
}

final class OtlpTimestampTests: XCTestCase {
    func testUnixNanosWithAndWithoutFractionalSeconds() {
        // 2026-06-13T10:00:00Z == 1781344800 epoch seconds.
        XCTAssertEqual(
            OtlpMapper.unixNanos(fromISO8601: "2026-06-13T10:00:00Z"),
            1_781_344_800_000_000_000)
        XCTAssertEqual(
            OtlpMapper.unixNanos(fromISO8601: "2026-06-13T10:00:00.123Z"),
            1_781_344_800_123_000_000)
        // Digit-exact down to nanoseconds (a Date roundtrip would round this).
        XCTAssertEqual(
            OtlpMapper.unixNanos(fromISO8601: "2026-06-13T10:00:00.123456789Z"),
            1_781_344_800_123_456_789)
        // Sub-nanosecond digits truncate.
        XCTAssertEqual(
            OtlpMapper.unixNanos(fromISO8601: "2026-06-13T10:00:00.1234567891Z"),
            1_781_344_800_123_456_789)
        // Timezone offsets after the fraction are honored (12:00+02:00 == 10:00Z).
        XCTAssertEqual(
            OtlpMapper.unixNanos(fromISO8601: "2026-06-13T12:00:00.5+02:00"),
            1_781_344_800_500_000_000)
    }

    func testUnixNanosRejectsMalformedAndPreEpoch() {
        XCTAssertNil(OtlpMapper.unixNanos(fromISO8601: ""))
        XCTAssertNil(OtlpMapper.unixNanos(fromISO8601: "not-a-date"))
        XCTAssertNil(OtlpMapper.unixNanos(fromISO8601: "2026-06-13"))  // date only
        XCTAssertNil(OtlpMapper.unixNanos(fromISO8601: "1969-12-31T23:59:59Z"))  // pre-epoch
    }
}

/// Golden tests for the authoritative ActivityEvent → log-record contract (the frozen downstream-SQL contract).
final class OtlpMapperTests: XCTestCase {
    private let context = OtlpMapper.SessionContext(
        sessionId: "s-test-1",
        traceId: "aaaa1111bbbb2222cccc3333dddd4444",
        spanId: "abcd1234abcd1234",
        startedAt: "2026-06-13T10:00:00.000Z",
        kind: "bdm-workshop",
        user: "petr@example.com",
        instanceName: "Padak's MacBook Pro",
        areaId: "area-fin",
        areaName: "Finance"
    )

    func testFullEventMapsEveryAttribute() {
        let event = ActivityEvent(
            sessionId: "s-test-1",
            eventId: "s-test-1-1",
            sequence: 1,
            timestamp: "2026-06-13T10:00:00.123Z",
            eventType: "click",
            url: "app://com.apple.finder",
            pageTitle: "Downloads",
            system: "Finder",
            target: EventTarget(
                tag: "AXButton", role: "AXButton", accessibleName: "Open", text: "Open",
                boundingBox: BoundingBox(x: 100.5, y: 200.25, width: 80, height: 24)),
            value: "Open",
            inputMasked: true,
            isSensitive: false,
            screenshotId: "84580000",
            labelId: "lbl-1",
            label: "sales commissions",
            processId: "proc-inv",
            process: "Invoicing"
        )
        let record = OtlpMapper.logRecord(for: event, in: context)

        XCTAssertEqual(record.body, .string("click"))
        XCTAssertEqual(record.severityText, "INFO")
        XCTAssertEqual(record.severityNumber, 9)
        XCTAssertEqual(record.traceId, context.traceId)
        XCTAssertEqual(record.spanId, context.spanId)
        // Record time = the event's own captured timestamp, not "now".
        XCTAssertEqual(record.timeUnixNano, "1781344800123000000")
        XCTAssertEqual(record.observedTimeUnixNano, record.timeUnixNano)

        // EXACT attribute set — any drift here breaks the downstream SQL contract.
        XCTAssertEqual(
            attributeDict(record.attributes),
            [
                "session.id": .string("s-test-1"),
                "sessionId": .string("s-test-1"),
                "eventId": .string("s-test-1-1"),
                "sequence": .int(1),
                "url": .string("app://com.apple.finder"),
                "page_title": .string("Downloads"),
                "system": .string("Finder"),
                "value": .string("Open"),
                "input_masked": .bool(true),
                "is_sensitive": .bool(false),
                "screenshot_id": .string("84580000"),
                "target.tag": .string("AXButton"),
                "target.role": .string("AXButton"),
                "target.accessibleName": .string("Open"),
                "target.selectorCandidates": .string(""),
                "target.boundingBox.x": .double(100.5),
                "target.boundingBox.y": .double(200.25),
                "target.boundingBox.width": .double(80),
                "target.boundingBox.height": .double(24),
                "enduser.id": .string("petr@example.com"),
                "host.name": .string("Padak's MacBook Pro"),
                "label.id": .string("lbl-1"),
                "label.name": .string("sales commissions"),
                "session.kind": .string("bdm-workshop"),
                "area.id": .string("area-fin"),
                "area.name": .string("Finance"),
                "process.id": .string("proc-inv"),
                "process.name": .string("Invoicing"),
            ])
    }

    func testMinimalEventCoercesStringsAndOmitsNumerics() {
        let event = ActivityEvent(
            sessionId: "s-test-1",
            eventId: "s-test-1-0",
            timestamp: "2026-06-13T10:00:00Z",
            eventType: "navigate",
            url: "app://x"
        )
        let record = OtlpMapper.logRecord(for: event, in: context)
        let attrs = attributeDict(record.attributes)

        // Strings: nil → "" (the key is ALWAYS present, value empty).
        XCTAssertEqual(
            attrs,
            [
                "session.id": .string("s-test-1"),
                "sessionId": .string("s-test-1"),
                "eventId": .string("s-test-1-0"),
                "url": .string("app://x"),
                "page_title": .string(""),
                "system": .string(""),
                "value": .string(""),
                "input_masked": .bool(false),
                "is_sensitive": .bool(false),
                "screenshot_id": .string(""),
                "target.tag": .string(""),
                "target.role": .string(""),
                "target.accessibleName": .string(""),
                "target.selectorCandidates": .string(""),
                "enduser.id": .string("petr@example.com"),
                "host.name": .string("Padak's MacBook Pro"),
                // No label open: keys ALWAYS present, value "" (never omitted, never null).
                "label.id": .string(""),
                "label.name": .string(""),
                "session.kind": .string("bdm-workshop"),
                // Area is session-scoped (from the context) so it rides even a minimal event; the
                // Process is segment-scoped and this event has none, so it is "" (present, empty).
                "area.id": .string("area-fin"),
                "area.name": .string("Finance"),
                "process.id": .string(""),
                "process.name": .string(""),
            ])
        // Numerics: the keys must be ABSENT (never ""), or Keboola column typing corrupts.
        for key in [
            "sequence",
            "target.boundingBox.x", "target.boundingBox.y",
            "target.boundingBox.width", "target.boundingBox.height",
            "viewport.width", "viewport.height", "viewport.scrollX", "viewport.scrollY",
        ] {
            XCTAssertNil(attrs[key], "numeric key \(key) must be omitted when absent")
        }
    }

    func testNarrationEventMapsToDedicatedRecord() {
        // Narration rides the spool as a synthetic ActivityEvent; the mapper gives it a
        // dedicated shape (audio ref + session start) PLUS enduser.id + host.name,
        // which ride every record so even narration is attributable to its user and machine
        // (without enduser.id the narration row would split the session under a NULL user).
        let event = ActivityEvent(
            sessionId: "s-test-1",
            eventId: "s-test-1-99",
            timestamp: "2026-06-13T10:05:00Z",
            eventType: "narration",
            url: "app://session",
            audioFileId: "84580000",
            labelId: "lbl-1",
            label: "sales commissions"
        )
        let record = OtlpMapper.logRecord(for: event, in: context)
        XCTAssertEqual(record.body, .string("narration"))
        XCTAssertEqual(
            attributeDict(record.attributes),
            [
                "session.id": .string("s-test-1"),
                "sessionId": .string("s-test-1"),
                "audio_file_id": .string("84580000"),
                "session.startedAt": .string("2026-06-13T10:00:00.000Z"),
                // Identity (WHO) + machine (WHICH) ride the narration record too.
                "enduser.id": .string("petr@example.com"),
                "host.name": .string("Padak's MacBook Pro"),
                // Audio ties to its label (mic records only while a label is open).
                "label.id": .string("lbl-1"),
                "label.name": .string("sales commissions"),
                "session.kind": .string("bdm-workshop"),
                // The narration record carries the session Area + (here unset) Process too.
                "area.id": .string("area-fin"),
                "area.name": .string("Finance"),
                "process.id": .string(""),
                "process.name": .string(""),
            ])
    }

    func testLabelStartEventCarriesLabelAttributes() {
        // A label_start event IS the bracketed-segment boundary: it carries the freshly-minted
        // labelId + human name, which then ride every subsequent event until the label ends.
        let event = ActivityEvent(
            sessionId: "s-test-1",
            eventId: "s-test-1-5",
            timestamp: "2026-06-13T10:02:00Z",
            eventType: "label_start",
            url: "app://session",
            labelId: "lbl-1",
            label: "sales commissions"
        )
        let record = OtlpMapper.logRecord(for: event, in: context)
        XCTAssertEqual(record.body, .string("label_start"))
        let attrs = attributeDict(record.attributes)
        XCTAssertEqual(attrs["label.id"], .string("lbl-1"))
        XCTAssertEqual(attrs["label.name"], .string("sales commissions"))
        // The explicit session type rides every record (the reliable per-event carrier).
        XCTAssertEqual(attrs["session.kind"], .string("bdm-workshop"))
    }

    func testUnparseableTimestampFallsBackToNow() {
        let event = ActivityEvent(
            sessionId: "s-test-1", eventId: "s-test-1-0",
            timestamp: "garbage", eventType: "click", url: "app://x"
        )
        let now = Date(timeIntervalSince1970: 1_781_344_800)
        let record = OtlpMapper.logRecord(for: event, in: context, now: now)
        XCTAssertEqual(record.timeUnixNano, "1781344800000000000")
    }

    func testSpanBuilderWithAndWithoutKind() {
        let span = OtlpMapper.span(in: context, endedAt: "2026-06-13T10:01:00Z")
        XCTAssertEqual(span.name, "capture-session")
        XCTAssertEqual(span.kind, 1)  // SPAN_KIND_INTERNAL
        XCTAssertEqual(span.traceId, context.traceId)
        XCTAssertEqual(span.spanId, context.spanId)
        XCTAssertEqual(span.startTimeUnixNano, "1781344800000000000")
        XCTAssertEqual(span.endTimeUnixNano, "1781344860000000000")
        XCTAssertEqual(
            attributeDict(span.attributes),
            [
                "session.id": .string("s-test-1"),
                "session.kind": .string("bdm-workshop"),
                "area.id": .string("area-fin"),
                "area.name": .string("Finance"),
                "session.endedAt": .string("2026-06-13T10:01:00Z"),
            ])

        // Normal captures carry no session.kind nor Area at all (keys absent, not "").
        var plain = context
        plain.kind = nil
        plain.areaId = nil
        let plainSpan = OtlpMapper.span(in: plain, endedAt: "2026-06-13T10:01:00Z")
        XCTAssertEqual(
            attributeDict(plainSpan.attributes),
            [
                "session.id": .string("s-test-1"),
                "session.endedAt": .string("2026-06-13T10:01:00Z"),
            ])
    }

    func testRequestEnvelopesCarryResourceAndScope() throws {
        let events = [
            ActivityEvent(
                sessionId: "s-test-1", eventId: "s-test-1-0",
                timestamp: "2026-06-13T10:00:00Z", eventType: "session_start", url: "app://x"),
            ActivityEvent(
                sessionId: "s-test-1", eventId: "s-test-1-1", sequence: 1,
                timestamp: "2026-06-13T10:00:01Z", eventType: "click", url: "app://x"),
        ]
        let logs = OtlpMapper.logsRequest(events: events, in: context)
        XCTAssertEqual(logs.resourceLogs.count, 1)
        // Resource identity rides logs AND traces: service.name + the user (twice) + host.name
        // (forward-compat fallback — the per-event host.name is what downstream actually reads).
        let expectedResource: [String: Otlp.AnyValue] = [
            "service.name": .string("jasnost-capture"),
            "service.instance.id": .string("petr@example.com"),
            "enduser.id": .string("petr@example.com"),
            "host.name": .string("Padak's MacBook Pro"),
        ]
        XCTAssertEqual(attributeDict(logs.resourceLogs[0].resource.attributes), expectedResource)
        XCTAssertEqual(logs.resourceLogs[0].scopeLogs[0].scope.name, "jasnost.agent")
        XCTAssertEqual(logs.resourceLogs[0].scopeLogs[0].logRecords.count, 2)

        let trace = OtlpMapper.traceRequest(in: context, endedAt: "2026-06-13T10:01:00Z")
        XCTAssertEqual(attributeDict(trace.resourceSpans[0].resource.attributes), expectedResource)
        XCTAssertEqual(trace.resourceSpans[0].scopeSpans[0].spans.count, 1)

        // The encoded JSON must roundtrip through the same models (wire-shape sanity).
        let decoded = try JSONDecoder().decode(
            Otlp.ExportLogsServiceRequest.self, from: JSONEncoder().encode(logs))
        XCTAssertEqual(decoded, logs)
    }
}

File: windows/Tests/JazzCaptureCoreTests/OtlpMapperTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using JazzCaptureCore;

namespace JazzCaptureCoreTests;

/// <summary>
/// Unit-level goldens for projection rules the shared conformance fixtures do not cover:
/// narration as a total replacement, numeric omission, the "now" timestamp fallback, and the
/// span's conditional attributes.
/// </summary>
public sealed class OtlpMapperTests
{
    private static readonly SessionContext FullContext = new(
        SessionId: "sess-ctx",
        TraceId: "0123456789abcdef0123456789abcdef",
        SpanId: "0123456789abcdef",
        StartedAt: "2026-07-02T09:00:00.000Z",
        Kind: "process-mapping",
        User: "ann@acme.com",
        InstanceName: "Ann's PC",
        AreaId: "finance",
        AreaName: "Finance");

    [Fact]
    public void NarrationReplacesTheAttributeSetEvenWhenTheEventCarriesGenericFields()
    {
        // A narration event that also carries target/sequence/url/value/application: every one of
        // those must be ignored, leaving exactly the 13-key narration shape.
        ActivityEvent narration = new()
        {
            SessionId = "sess-conf-003",
            EventId = "evt-003",
            Sequence = 42,
            Timestamp = "2026-07-02T10:03:20.000Z",
            EventType = "narration",
            Url = "app://com.microsoft.Edge",
            Value = "ignored",
            System = "ignored",
            ScreenshotId = "shot-1",
            ClickCount = 2,
            DragEnd = new DragEnd(1.0, 2.0),
            InputMasked = true,
            IsSensitive = true,
            Application = new ApplicationRef("windows.aumid", "com.acme.app", "Acme", "1.0"),
            Target = new EventTarget
            {
                Tag = "Button",
                Role = "button",
                AccessibleName = "Save",
                BoundingBox = new BoundingBox(1, 2, 3, 4),
            },
            AudioFileId = "audio-77",
            LabelId = "lbl-9",
            Label = "Explain the refund flow",
            ProcessId = "refund",
            Process = "Refund handling",
        };

        IReadOnlyList<OtlpKeyValue> attributes = OtlpMapper.Attributes(narration, FullContext);

        Assert.Equal(
            new[]
            {
                "session.id",
                "sessionId",
                "audio_file_id",
                "session.startedAt",
                "enduser.id",
                "host.name",
                "label.id",
                "label.name",
                "session.kind",
                "area.id",
                "area.name",
                "process.id",
                "process.name",
            },
            attributes.Select(attribute => attribute.Key).ToArray());
        Assert.Equal("2026-07-02T09:00:00.000Z", StringAttribute(attributes, "session.startedAt"));
        Assert.Equal("audio-77", StringAttribute(attributes, "audio_file_id"));
    }

    [Fact]
    public void NarrationKeepsEventBodyAndTimestamp()
    {
        ActivityEvent narration = new()
        {
            SessionId = "sess-conf-003",
            EventId = "evt-003",
            Timestamp = "2026-07-02T10:03:20.000Z",
            EventType = "narration",
            Url = string.Empty,
        };

        JsonObject record = FirstLogRecord(OtlpMapper.LogsRequest(new[] { narration }, FullContext));

        Assert.Equal("narration", record["body"]!["stringValue"]!.GetValue<string>());
        Assert.Equal("1782986600000000000", record["timeUnixNano"]!.GetValue<string>());
        Assert.Equal("1782986600000000000", record["observedTimeUnixNano"]!.GetValue<string>());
    }

    [Fact]
    public void GenericAttributesOmitNumericKeysWhenTheValuesAreAbsent()
    {
        ActivityEvent minimal = new()
        {
            SessionId = "sess-conf-001",
            EventId = "evt-legacy-001",
            Timestamp = "2026-07-02T09:00:05.250Z",
            EventType = "click",
            Url = "app://com.google.Chrome",
        };

        IReadOnlyList<OtlpKeyValue> attributes = OtlpMapper.Attributes(minimal, FullContext);
        string[] keys = attributes.Select(attribute => attribute.Key).ToArray();

        Assert.Equal(31, keys.Length);
        Assert.DoesNotContain("sequence", keys);
        Assert.DoesNotContain("click_count", keys);
        Assert.DoesNotContain("drag_end.x", keys);
        Assert.DoesNotContain("drag_end.y", keys);
        Assert.DoesNotContain("target.boundingBox.x", keys);
        Assert.DoesNotContain("target.boundingBox.y", keys);
        Assert.DoesNotContain("target.boundingBox.width", keys);
        Assert.DoesNotContain("target.boundingBox.height", keys);

        // Strings coerce to "" and keep their key; booleans default to false and keep their key.
        Assert.Equal(string.Empty, StringAttribute(attributes, "page_title"));
        Assert.Equal(string.Empty, StringAttribute(attributes, "target.selectorCandidates"));
        Assert.False(attributes.Single(attribute => attribute.Key == "input_masked").Value.Flag);
        Assert.False(attributes.Single(attribute => attribute.Key == "is_sensitive").Value.Flag);
    }

    [Fact]
    public void GenericAttributesEmitNumericKeysWithProtoJsonTyping()
    {
        ActivityEvent rich = new()
        {
            SessionId = "sess-conf-001",
            EventId = "evt-drag-001",
            Sequence = 7,
            Timestamp = "2026-07-02T09:00:05.250Z",
            EventType = "drag",
            Url = "app://com.google.Chrome",
            ClickCount = 2,
            DragEnd = new DragEnd(420.5, 310.25),
            Target = new EventTarget { BoundingBox = new BoundingBox(20, 30, 180, 24) },
        };

        JsonObject record = FirstLogRecord(OtlpMapper.LogsRequest(new[] { rich }, FullContext));
        JsonArray attributes = record["attributes"]!.AsArray();

        Assert.Equal(39, attributes.Count);

        // int64 travels as a decimal STRING in proto3 JSON; doubles stay JSON numbers.
        JsonNode sequence = AttributeValue(attributes, "sequence");
        Assert.Equal(JsonValueKind.String, sequence["intValue"]!.GetValueKind());
        Assert.Equal("7", sequence["intValue"]!.GetValue<string>());
        Assert.Equal("2", AttributeValue(attributes, "click_count")["intValue"]!.GetValue<string>());

        JsonNode dragX = AttributeValue(attributes, "drag_end.x");
        Assert.Equal(JsonValueKind.Number, dragX["doubleValue"]!.GetValueKind());
        Assert.Equal(420.5, dragX["doubleValue"]!.GetValue<double>());
        Assert.Equal(24.0, AttributeValue(attributes, "target.boundingBox.height")["doubleValue"]!.GetValue<double>());

        // severityNumber is a bare number, the nano timestamps are strings.
        Assert.Equal(JsonValueKind.Number, record["severityNumber"]!.GetValueKind());
        Assert.Equal(JsonValueKind.String, record["timeUnixNano"]!.GetValueKind());
        Assert.Equal(JsonValueKind.False, AttributeValue(attributes, "input_masked")["boolValue"]!.GetValueKind());
    }

    [Fact]
    public void UnparseableEventTimestampFallsBackToTheInjectedNow()
    {
        ActivityEvent broken = new()
        {
            SessionId = "sess-conf-001",
            EventId = "evt-broken",
            Timestamp = "not-a-date",
            EventType = "click",
            Url = "app://com.google.Chrome",
        };
        DateTimeOffset now = DateTimeOffset.FromUnixTimeSeconds(1781344800);

        JsonObject record = FirstLogRecord(OtlpMapper.LogsRequest(new[] { broken }, FullContext, () => now));

        Assert.Equal("1781344800000000000", record["timeUnixNano"]!.GetValue<string>());
        Assert.Equal("1781344800000000000", record["observedTimeUnixNano"]!.GetValue<string>());
    }

    [Fact]
    public void UnparseableSpanTimestampsFallBackToTheInjectedNow()
    {
        SessionContext context = FullContext with { StartedAt = "2026-06-13" };
        DateTimeOffset now = DateTimeOffset.FromUnixTimeSeconds(1781344800);

        JsonObject span = FirstSpan(OtlpMapper.TraceRequest(context, "still-not-a-date", () => now));

        Assert.Equal("1781344800000000000", span["startTimeUnixNano"]!.GetValue<string>());
        Assert.Equal("1781344800000000000", span["endTimeUnixNano"]!.GetValue<string>());
        Assert.Equal("still-not-a-date", SpanAttribute(span, "session.endedAt"));
    }

    [Fact]
    public void SpanDropsUnsetKindAndAreaInsteadOfEmittingEmptyStrings()
    {
        SessionContext bare = FullContext with { Kind = null, AreaId = null, AreaName = null };

        JsonObject span = FirstSpan(OtlpMapper.TraceRequest(bare, "2026-07-02T09:12:00.000Z"));

        Assert.Equal(
            new[] { "session.id", "session.endedAt" },
            span["attributes"]!.AsArray().Select(attribute => attribute!["key"]!.GetValue<string>()).ToArray());
        // The span's session.id comes from the CONTEXT, unlike the log record's.
        Assert.Equal("sess-ctx", SpanAttribute(span, "session.id"));
        Assert.Equal(JsonValueKind.Number, span["kind"]!.GetValueKind());
        Assert.Equal(1, span["kind"]!.GetValue<int>());
    }

    [Fact]
    public void SpanEmitsAreaNameOnlyWhenAreaIdIsSet()
    {
        SessionContext orphanName = FullContext with { AreaId = null, AreaName = "Finance" };

        JsonObject span = FirstSpan(OtlpMapper.TraceRequest(orphanName, "2026-07-02T09:12:00.000Z"));
        string[] keys = span["attributes"]!.AsArray()
            .Select(attribute => attribute!["key"]!.GetValue<string>())
            .ToArray();

        Assert.DoesNotContain("area.id", keys);
        Assert.DoesNotContain("area.name", keys);

        SessionContext idOnly = FullContext with { AreaName = null };
        string[] idOnlyKeys = FirstSpan(OtlpMapper.TraceRequest(idOnly, "2026-07-02T09:12:00.000Z"))["attributes"]!
            .AsArray()
            .Select(attribute => attribute!["key"]!.GetValue<string>())
            .ToArray();

        Assert.Contains("area.id", idOnlyKeys);
        Assert.DoesNotContain("area.name", idOnlyKeys);
    }

    [Fact]
    public void LogRecordIdentityComesFromTheEventWhileSessionScopeComesFromTheContext()
    {
        ActivityEvent renamed = new()
        {
            SessionId = "sess-from-event",
            EventId = "evt-1",
            Timestamp = "2026-07-02T09:00:05.250Z",
            EventType = "click",
            Url = "app://x",
            Application = new ApplicationRef("windows.aumid", "com.acme.app", "Acme", "1.0"),
            DocumentUrl = "file:///doc",
            PageTitle = "Title",
            SelectedText = "sel",
            ClipboardText = "clip",
            Label = "Label name",
            LabelId = "lbl-1",
            Process = "Process name",
            ProcessId = "proc-1",
            Target = new EventTarget { Text = "never emitted" },
        };

        IReadOnlyList<OtlpKeyValue> attributes = OtlpMapper.Attributes(renamed, FullContext);

        Assert.Equal("sess-from-event", StringAttribute(attributes, "session.id"));
        Assert.Equal("sess-from-event", StringAttribute(attributes, "sessionId"));
        Assert.Equal("ann@acme.com", StringAttribute(attributes, "enduser.id"));
        Assert.Equal("Ann's PC", StringAttribute(attributes, "host.name"));
        // Key renames from the digest table.
        Assert.Equal("com.acme.app", StringAttribute(attributes, "application.id"));
        Assert.Equal("file:///doc", StringAttribute(attributes, "document.url"));
        Assert.Equal("Title", StringAttribute(attributes, "page_title"));
        Assert.Equal("sel", StringAttribute(attributes, "selected_text"));
        Assert.Equal("clip", StringAttribute(attributes, "clipboard_text"));
        Assert.Equal("Label name", StringAttribute(attributes, "label.name"));
        Assert.Equal("Process name", StringAttribute(attributes, "process.name"));
        // target.text is never projected.
        Assert.DoesNotContain("target.text", attributes.Select(attribute => attribute.Key));
    }

    [Fact]
    public void EmptyEventListStillEmitsTheSingleResourceAndScopeWrapper()
    {
        JsonObject request = OtlpMapper.LogsRequest(Array.Empty<ActivityEvent>(), FullContext);

        JsonArray resourceLogs = request["resourceLogs"]!.AsArray();
        Assert.Single(resourceLogs);
        JsonArray scopeLogs = resourceLogs[0]!["scopeLogs"]!.AsArray();
        Assert.Single(scopeLogs);
        Assert.Equal("jazz.agent", scopeLogs[0]!["scope"]!["name"]!.GetValue<string>());
        Assert.Empty(scopeLogs[0]!["logRecords"]!.AsArray());
        Assert.Equal(
            new[] { "service.name", "service.instance.id", "enduser.id", "host.name" },
            resourceLogs[0]!["resource"]!["attributes"]!.AsArray()
                .Select(attribute => attribute!["key"]!.GetValue<string>())
                .ToArray());
    }

    [Fact]
    public void UnknownEventPropertiesAreIgnoredOnDeserialization()
    {
        const string json = """
            {
              "sessionId": "s",
              "eventId": "e",
              "timestamp": "2026-07-02T09:00:05.250Z",
              "eventType": "click",
              "url": "app://x",
              "tabId": 12,
              "frameId": 0,
              "rrwebChunkId": "chunk",
              "viewport": { "width": 100, "height": 200, "scrollX": 0, "scrollY": 0 },
              "target": { "selectorCandidates": [ { "kind": "css", "value": "#a" } ], "attributes": { "id": "a" } }
            }
            """;

        ActivityEvent decoded = JsonSerializer.Deserialize<ActivityEvent>(json)!;

        Assert.Equal("click", decoded.EventType);
        Assert.NotNull(decoded.Target);
        Assert.Null(decoded.Sequence);
    }

    private static JsonObject FirstLogRecord(JsonObject logsRequest) =>
        logsRequest["resourceLogs"]![0]!["scopeLogs"]![0]!["logRecords"]![0]!.AsObject();

    private static JsonObject FirstSpan(JsonObject traceRequest) =>
        traceRequest["resourceSpans"]![0]!["scopeSpans"]![0]!["spans"]![0]!.AsObject();

    private static string StringAttribute(IReadOnlyList<OtlpKeyValue> attributes, string key) =>
        attributes.Single(attribute => attribute.Key == key).Value.Text;

    private static string SpanAttribute(JsonObject span, string key) =>
        AttributeValue(span["attributes"]!.AsArray(), key)["stringValue"]!.GetValue<string>();

    private static JsonNode AttributeValue(JsonArray attributes, string key) =>
        attributes.Single(attribute => attribute!["key"]!.GetValue<string>() == key)!["value"]!;
}
